File: src/matrix.js
/*
 * This file is adapted from the Python attack_flow.matrix module so that we can generate similar Navigator
 * overlay graphics from the browser instead of the command line.
 */
const TACTIC_CLASS_RE = /tactic ([-a-z]+)/;
const TECHNIQUE_CLASS_RE = /(sub)?technique (T[.0-9]+)/;
const TRANSLATE_RE = /^translate\(([^,]+),\s*([^,]+)\)$/;
const OVERLAY_COLOR = "rgba(204, 0, 0, 0.75)";
const ELLIPSE_SIZE_MULTIPLIER = 0.6;
const LINE_BENDINESS = 0.15;
const ARROWHEAD_WIDTH = 7;
const ARROWHEAD_HEIGHT = 5;
const ARROW_OFFSET = 30;
const STROKE_WIDTH = 2;
const SHOW_CONTROL_POINTS = false; // for debugging, set to true

// Map tactic IDs to the CSS class that Navigator emits.
const TACTIC_LOOKUP = {
    // Enterprise
    "TA0043": "reconnaissance",
    "TA0042": "resource-development",
    "TA0001": "initial-access",
    "TA0002": "execution",
    "TA0003": "persistence",
    "TA0004": "privilege-sscalation",
    "TA0005": "defense-evasion",
    "TA0006": "credential-access",
    "TA0007": "discovery",
    "TA0008": "lateral-movement",
    "TA0009": "collection",
    "TA0011": "command-and-control",
    "TA0010": "exfiltration",
    "TA0040": "impact",
    // Mobile:
    "TA0027": "initial-access",
    "TA0041": "execution",
    "TA0028": "persistence",
    "TA0029": "privilege-escalation",
    "TA0030": "defense-evasion",
    "TA0031": "credential-access",
    "TA0032": "discovery",
    "TA0033": "lateral-movement",
    "TA0035": "collection",
    "TA0037": "command-and-control",
    "TA0036": "exfiltration",
    "TA0034": "impact",
    "TA0038": "network-effects",
    "TA0039": "remote-service-effects",
    // ICS:
    "TA0108": "initial-access",
    "TA0104": "execution",
    "TA0110": "persistence",
    "TA0111": "privilege-sscalation",
    "TA0103": "evasion",
    "TA0102": "discovery",
    "TA0109": "lateral-movement",
    "TA0100": "collection",
    "TA0101": "command-and-control",
    "TA0107": "inhibit-response-function",
    "TA0106": "impair-process-control",
    "TA0105": "impact",
}

/**
 * Render an SVG with the flow drawn on top of a given ATT&CK Navigator layer.
 * @param {*} svgXml
 * @param {*} flowBundle
 */
function render(svgXml, flowStix) {
    const domParser = new DOMParser();
    const svgDoc = domParser.parseFromString(svgXml, "application/xml");

    // Build a lookup table from technique ID -> translation coordinates for that technique's <g> element.
    const techniqueGeometries = {};
    _enumerateTechniqueGeometries(techniqueGeometries, svgDoc.children[0]);

    // Create a new <g> to hold all of the Attack Flow overlay elements.
    const attackFlowOverlay = document.createElement("g");
    attackFlowOverlay.setAttribute("class", "attack-flow-overlay");
    attackFlowOverlay.appendChild(_getArrowheadMarker());
    svgDoc.children[0].appendChild(attackFlowOverlay);

    // Load the Attack Flow STIX bundle and convert it to an action graph.
    const flowGraph = _convertBundleToGraph(flowStix);
    const actionGraph = _induceActionGraph(flowGraph);

    // Draw ellipses around the (sub-)techniques.
    const stack = [...Object.values(actionGraph.children)];

    while (node = stack.shift()) {
        const tid = node["technique_id"];
        const translation = _lookupGeometry(node, techniqueGeometries);
        if (translation) {
            const techniqueOverlay = _createTechniqueOverlay(tid, translation);
            attackFlowOverlay.appendChild(techniqueOverlay);
        }
        stack.unshift(...Object.values(node.children));
    }

    // Draw arrows from each technique to its children. Note that we are making a separate pass over the nodes
    // because in the previous loop we might change a subtechnique to its parent if the subtechnique is not in
    // the SVG, and we can't draw any arrows until we've confirmed which TIDs to use.
    stack.push(...Object.values(actionGraph.children));

    while (node = stack.shift()) {
        const sourceTid = node["technique_id"];
        const sourceGeometry = _lookupGeometry(node, techniqueGeometries);
        for (const child of Object.values(node.children)) {
            const targetTid = child["technique_id"];
            const targetGeometry = _lookupGeometry(child, techniqueGeometries);
            if (!(sourceGeometry && targetGeometry)) {
                // This could happen if a specified technique does not exist in this matrix, e.g. using mobile
                // techniques on an enterprise matrix.
                continue;
            }
            const arrowOverlay = _createRelationshipOverlay(
                sourceTid,
                targetTid,
                sourceGeometry,
                targetGeometry,
                SHOW_CONTROL_POINTS);
            attackFlowOverlay.appendChild(arrowOverlay);
        }
        stack.push(...Object.values(node.children));

    }

    // The SVG doesn't render correctly if we directly attach it to the page. Instead, create a new SVG
    // element and copy the contents into it.
    const svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
    svg.setAttribute("width", svgDoc.documentElement.getAttribute("width"));
    svg.setAttribute("height", svgDoc.documentElement.getAttribute("height"));
    svg.innerHTML = svgDoc.documentElement.innerHTML;
    document.querySelector("#preview").appendChild(svg);
}

/**
 * Convert an Attack Flow STIX bundle to a graph representation.
 * @param {*} bundleSrc
 * @returns
 */
function _convertBundleToGraph(bundleSrc) {
    const bundle = JSON.parse(bundleSrc);
    let flowMeta = null;
    const objectLookup = {};

    // Find the flow object.
    for (const obj of bundle.objects) {
        if (obj.type === "attack-flow") {
            flowMeta = obj;
        }
        objectLookup[obj.id] = obj;
        obj.children = {};
    }

    if (!flowMeta) {
        throw new Error("Cannot find an attack-flow object in the bundle.")
    }

    // The initial graph contains the start ref nodes.
    const root = { type: "root", children: {} };
    const stack = [];
    for (const startRef of flowMeta["start_refs"]) {
        stack.unshift([root, objectLookup[startRef]]);
    }

    // Add children from direct refs.
    let i = 0;
    while (item = stack.shift()) {
        const [parent, node] = item;
        parent.children[node.id] = node;
        for (const [key, value] of Object.entries(node)) {
            if (key.substring(key.length - 4) == "_ref") {
                stack.unshift([node, objectLookup[value]]);
            } else if (key.substring(key.length - 5) == "_refs") {
                for (const ref of value) {
                    stack.unshift([node, objectLookup[ref]]);
                }
            }
        }
    }

    // Add children from relationships.
    for (const rel of bundle.objects) {
        if (rel.type == "relationship") {
            const src = objectLookup[rel["source_ref"]];
            const target = objectLookup[rel["target_ref"]];
            src.children[target.id] = target;
        }
    }

    return root;
}

/**
 * Given an Attack Flow graph, return the induced graph containing only attack-action nodes.
 * @param {*} flowGraph
 * @returns
 */
function _induceActionGraph(flowGraph) {
    // Deep copy the flow graph.
    const actionGraph = JSON.parse(JSON.stringify(flowGraph));
    const stack = [];

    // Start with the start ref nodes.
    for (const [id, child] of Object.entries(actionGraph.children)) {
        stack.unshift([actionGraph, child]);
    }

    // Iterate through nodes and remove the ones that are not actions.
    while (item = stack.shift()) {
        const [parent, node] = item;

        if (node.type != "attack-action") {
            delete parent.children[node.id];
            Object.assign(parent.children, node.children);
            for (const [id, child] of Object.entries(node.children)) {
                stack.unshift([parent, child]);
            }
        } else {
            for (const [id, child] of Object.entries(node.children)) {
                stack.unshift([node, child]);
            }
        }
    }

    return actionGraph;
}

/**
 * Iterate over an XML tree and create objects that represent the geometry for each <g> element for a
 * technique.
 *
 * SVG doesn't have any concept of depth or z-index. It draws all items in document order. So in order to draw
 * an overlay, the overlay must come after all elements that it intersects. The straightforward way to do this
 * is to append the element to the end of the document, but then it won't be translated properly since it's no
 * longer nested inside the same <g> as the target element. So this method traverses the tree and keeps track
 * of the sum of translations for each element, then writes a new technique <g> back to techniqueGeometries
 * with the technique ID as key and its translation as values.
 * @param {*} techniqueGeometries - this object is edited in place!
 * @param {*} node
 * @param {*} parentX
 * @param {*} parentY
 */
function _enumerateTechniqueGeometries(techniqueGeometries, node, parentX = 0, parentY = 0, tactic = null) {
    for (const child of node.children) {
        let childX = parentX;
        let childY = parentY;
        let tempTactic = tactic;

        if (transform = child.getAttribute("transform")) {
            if (match = transform.match(TRANSLATE_RE)) {
                childX += parseFloat(match[1]);
                childY += parseFloat(match[2]);
            } else {
                throw new Exception(`Unknown SVG transform: "${transform}" (was this SVG created by ATT&CK Navigator?)`);
            }
        }

        const childClass = child.getAttribute("class");
        if (childClass && (match = childClass.match(TACTIC_CLASS_RE))) {
            // If we traverse a tactic node (which is a <g>) then pass the tactic name down to the recursive
            // calls.
            tempTactic = match[1];
        } else if (childClass && (match = childClass.match(TECHNIQUE_CLASS_RE))) {
            const techniqueId = match[2];
            try {
                const rect = child.querySelector("rect");
                const width = parseFloat(rect.getAttribute("width"));
                const height = parseFloat(rect.getAttribute("height"));
                // Store geometries twice: once under TechniqueID and once more under a compound key
                // Tactic-TechniqueID. The renderer will use Tactic if available and fall back to just
                // TechniqueID.
                const geom = { x: childX, y: childY, width, height };
                techniqueGeometries[techniqueId] = geom;
                if (tempTactic) {
                    techniqueGeometries[`${tempTactic}-${techniqueId}`] = geom;
                }
            } catch (e) {
                console.log(`Could not find a <rect> with width and height attributes for technique ${techniqueId}`);
            }
        }

        _enumerateTechniqueGeometries(techniqueGeometries, child, childX, childY, tempTactic);
    }
}

/**
 * Look up the geometry for a given node (i.e. where it is located in the graphic). If the node contains a
 * tactic ID, then we attempt to look up the node within that tactic column, otherwise we return an arbitrary
 * geometry matching that node's technique ID.
 * @param {*} node
 * @param {*} techniqueGeometries
 */
function _lookupGeometry(node, techniqueGeometries) {
    const techniqueId = node["technique_id"];
    const tacticId = node["tactic_id"];
    const tactic = TACTIC_LOOKUP[tacticId];
    let geometry;

    // Try looking up by tactic + technique ID first.
    if (tactic) {
        geometry = techniqueGeometries[`${tactic}-${techniqueId}`];
    }

    // Fall back to looking up by technique ID alone.
    if (!geometry) {
        geometry = techniqueGeometries[techniqueId];
    }

    // Finally, check if its a subtechnique and its prent technique can be looked up.
    if (!geometry) {
        if (techniqueId.includes(".")) {
            // If the TID is not found and it's a subtechnique, then check if the parent TID exists.
            const parentTid = techniqueId.split(".")[0];

            // Look up by tactic + technique ID first.
            if (tactic) {
                geometry = techniqueGeometries[`${tactic}-${parentTid}`];
            }

            // Fall back to looking up by technique ID alone.
            if (!geometry) {
                geometry = techniqueGeometries[parentTid];
            }

            if (geometry) {
                console.warn(`Did not find subtechnique "${techniqueId}" in the SVG; substituting its parent technique "${parentTid}".`)
                // Hack: update the technique ID so we can draw the arrows pointing at it later.
                node["technique_id"] = parentTid;
            } else {
                console.error(`Did not find subtechnique "${techniqueId}" or its parent technique "${parentTid}" in the SVG.`);
            }
        } else {
            console.error(`Did not find TID=${techniqueId} in the SVG.`);
        }
    }

    return geometry;
}

/**
 * Create an ellipse object that goes around the specified technique.
 * @param {*} techniqueId
 * @param {*} techniqueGeometry
 * @returns
 */
function _createTechniqueOverlay(techniqueId, techniqueGeometry) {
    const { x, y, width, height } = techniqueGeometry;
    const overlay = document.createElement("g");
    overlay.classList.add("overlay", techniqueId);
    overlay.setAttribute("transform", `translate(${x}, ${y})`);

    const ellipse = document.createElement("ellipse");
    ellipse.setAttribute("rx", (width * ELLIPSE_SIZE_MULTIPLIER).toString());
    ellipse.setAttribute("ry", (height * ELLIPSE_SIZE_MULTIPLIER).toString());
    ellipse.setAttribute("cx", (width / 2).toString());
    ellipse.setAttribute("cy", (height / 2).toString());
    ellipse.setAttribute("fill", "none");
    ellipse.setAttribute("stroke", OVERLAY_COLOR);
    ellipse.setAttribute("strokeWidth", STROKE_WIDTH);
    overlay.appendChild(ellipse);

    return overlay;
}

/**
 * Draw an arrow from one technique to another.
 * @param {*} sourceTid
 * @param {*} targetTid
 * @param {*} sourceGeometry
 * @param {*} targetGeometry
 * @param {*} showControlPoints
 */
function _createRelationshipOverlay(sourceTid, targetTid, sourceGeometry, targetGeometry, showControlPoints) {
    // Create a container element for the arrow.
    const overlay = document.createElement("g");
    overlay.classList.add("overlay", `${sourceTid}-${targetTid}`);
    const translateX = sourceGeometry.x + sourceGeometry.width / 2;
    const translateY = sourceGeometry.y + sourceGeometry.height / 2;
    overlay.setAttribute("transform", `translate(${translateX}, ${translateY})`);

    // Compute the coordinates for the path.
    const startRX = sourceGeometry.width * ELLIPSE_SIZE_MULTIPLIER;
    const startRY = sourceGeometry.height * ELLIPSE_SIZE_MULTIPLIER;
    let startX = ARROW_OFFSET;
    let startY = Math.sqrt(startRY ** 2 * (1 - startX ** 2 / startRX ** 2));

    const targetRX = targetGeometry.width * ELLIPSE_SIZE_MULTIPLIER;
    const targetRY = targetGeometry.height * ELLIPSE_SIZE_MULTIPLIER;
    const targetDX = ARROW_OFFSET;
    const targetDY = Math.sqrt(targetRY ** 2 * (1 - targetDX ** 2 / targetRX ** 2));
    let targetX = targetGeometry.x - sourceGeometry.x;
    let targetY = targetGeometry.y - sourceGeometry.y;

    // Make room for the arrow head. The slope is the same as control_dy / control_dx so that the arrow points
    // in the same direction as the line end.
    distance = Math.sqrt((targetX - startX) ** 2 + (targetY - startY) ** 2);
    const controlDX = distance * 0.1;
    const controlDY = distance * 0.2;
    arrowDX = Math.sqrt((ARROWHEAD_WIDTH * 2) ** 2 / 5)
    arrowDY = arrowDX * 2
    let control1X, control1Y, control2X, control2Y;

    if (targetGeometry.x < sourceGeometry.x) {
        startX = -startX
        targetX += targetDX + arrowDX
        control1X = startX - controlDX
        control2X = targetX + controlDX
    } else {
        targetX -= targetDX + arrowDX
        control1X = startX + controlDX
        control2X = targetX - controlDX
    }
    if (targetGeometry.y < sourceGeometry.y) {
        startY = -startY
        targetY += targetDY + arrowDY
        control1Y = startY - controlDY
        control2Y = targetY + controlDY
    } else {
        targetY -= targetDY + arrowDY
        control1Y = startY + controlDY
        control2Y = targetY - controlDY
    }

    // Show control points, if requested for debugging.
    if (showControlPoints) {
        startPoint = document.createElement("circle");
        startPoint.setAttribute("cx", startX);
        startPoint.setAttribute("cy", startY);
        startPoint.setAttribute("r", 2);
        startPoint.style.fill = "green";
        overlay.appendChild(startPoint);

        targetPoint = document.createElement("circle");
        targetPoint.setAttribute("cx", targetX);
        targetPoint.setAttribute("cy", targetY);
        targetPoint.setAttribute("r", 2);
        targetPoint.style.fill = "blue";
        overlay.appendChild(targetPoint);

        controlPoint1 = document.createElement("circle");
        controlPoint1.setAttribute("cx", control1X);
        controlPoint1.setAttribute("cy", control1Y);
        controlPoint1.setAttribute("r", 2);
        controlPoint1.style.fill = "red";
        overlay.appendChild(controlPoint1);

        controlPoint2 = document.createElement("circle");
        controlPoint2.setAttribute("cx", control2X);
        controlPoint2.setAttribute("cy", control2Y);
        controlPoint2.setAttribute("r", 2);
        controlPoint2.style.fill = "orange";
        overlay.appendChild(controlPoint2);
    }

    // Create the path.
    path = document.createElement("path");
    path.setAttribute("d", `M ${startX} ${startY} C ${control1X} ${control1Y} ${control2X} ${control2Y} ${targetX} ${targetY}`);
    path.setAttribute("marker-end", "url(#attack-flow-arrowhead)");
    path.style.stroke = OVERLAY_COLOR;
    path.style.strokeWidth = STROKE_WIDTH;
    path.style.fill = "none";
    overlay.append(path)

    return overlay;
}

/**
 * Create an arrowhead element that can be re-used for each path
 * @returns
 */
function _getArrowheadMarker() {
    const marker = document.createElement("marker");
    marker.setAttribute("id", "attack-flow-arrowhead");
    marker.setAttribute("markerWidth", ARROWHEAD_WIDTH);
    marker.setAttribute("markerHeight", ARROWHEAD_HEIGHT);
    marker.setAttribute("refX", "0");
    marker.setAttribute("refY", ARROWHEAD_HEIGHT / 2);
    marker.setAttribute("orient", "auto");

    const poly = document.createElement("polygon");
    const halfHeight = ARROWHEAD_HEIGHT / 2;
    poly.setAttribute("points", `0 0, ${ARROWHEAD_WIDTH} ${halfHeight}, 0 ${ARROWHEAD_HEIGHT}`);
    poly.style.fill = OVERLAY_COLOR;
    marker.appendChild(poly);

    return marker
}

File: src/attack_flow/matrix.py
"""
Code for drawing an Attack Flow action graph on top of an ATT&CK Navigator SVG.

This is quite brittle since it depends on the internal structure of the Navigator SVG. For future proofing we
might want to read the Navigator layer and generate our own rendering of the matrix, but in the interest of
time this simpler approach will suffice.

This code has been ported to JavaScript for use on the Attack Flow website. That version of the code contains
some improvements, such as the ability to incorporate tactic information when selecting a cell on the matrix.
"""

from collections import namedtuple
import logging
import math
import re
from xml.etree import ElementTree

from defusedxml.ElementTree import parse as defusedxml_parse

import attack_flow.graph


logger = logging.getLogger(__name__)
TECHNIQUE_CLASS_RE = re.compile(r"(sub)?technique (T[.0-9]+)")
TRANSLATE_RE = re.compile(r"^translate\(([^,]+),\s*([^,]+)\)$")
OVERLAY_COLOR = "rgba(204, 0, 0, 0.75)"
ELLIPSE_SIZE_MULTIPLIER = 0.6
LINE_BENDINESS = 0.15
ARROWHEAD_WIDTH = 7
ARROWHEAD_HEIGHT = 5
ARROW_OFFSET = 30
STROKE_WIDTH = 2
_TechniqueGeometry = namedtuple("_TechniqueGeometry", ["x", "y", "width", "height"])


def render(matrix_file, flow_bundle, out_file, show_control_points=False):
    """
    Render a flow on top of an ATT&CK matrix SVG.

    :param file matrix_file: A file-like object containing the SVG content.
    :param stix2.Bundle flow_bundle: The flow as a STIX bundle.
    :param file out_file: A file-like object to write the result to.
    :param show_control_points: For debugging, display the control points for the
        arrows' curves.
    """
    ElementTree.register_namespace("", "http://www.w3.org/2000/svg")
    tree = defusedxml_parse(matrix_file)

    # Build a lookup table from technique ID -> translation coordinates for that
    # technique's <g> element.
    technique_geometries = dict()
    _enumerate_technique_geometries(technique_geometries, tree.getroot())

    # Create a new <g> to hold all of the Attack Flow overlay elements.
    attack_flow_overlay = ElementTree.Element("g", {"class": "attack-flow-overlay"})
    attack_flow_overlay.append(_get_arrowhead_marker())
    tree.getroot().append(attack_flow_overlay)

    # Create the Attack Flow overlay elements.
    full_graph = attack_flow.graph.bundle_to_networkx(flow_bundle)
    graph = attack_flow.graph.induce_action_graph(full_graph)

    for node, data in graph.nodes(data=True):
        try:
            tid = data["technique_id"]
        except KeyError:
            logger.warning("Node (%s) does not have a technique ID.", node)
            continue
        if translation := technique_geometries.get(tid):
            technique_overlay = _create_technique_overlay(tid, translation)
            attack_flow_overlay.append(technique_overlay)
        else:
            if "." in tid:
                tid2, _ = tid.split(".")
                if translation := technique_geometries.get(tid2):
                    logger.warning(
                        "Did not find subtechnique %s; replacing with parent technique %s.",
                        tid,
                        tid2,
                    )
                    data["technique_id"] = tid2
                    technique_overlay = _create_technique_overlay(tid2, translation)
                    attack_flow_overlay.append(technique_overlay)
                else:
                    logger.warning(
                        "Did not find subtechnique %s or parent technique %s in input SVG.",
                        tid,
                        tid2,
                    )
            else:
                logger.warning(f"Did not find technique ID {tid} in input SVG.")

    for src_id, target_id in graph.edges:
        src_data = graph.nodes[src_id]
        target_data = graph.nodes[target_id]
        try:
            src_tid = src_data["technique_id"]
            target_tid = target_data["technique_id"]
            src_geom = technique_geometries[src_tid]
            target_geom = technique_geometries[target_tid]
        except KeyError:
            # Note that missing technique IDs are logged above.
            continue
        arrow_overlay = _create_relationship_overlay(
            src_tid, src_geom, target_tid, target_geom, show_control_points
        )
        attack_flow_overlay.append(arrow_overlay)

    # Write the SVG to output.
    tree.write(out_file, xml_declaration=True)


def _enumerate_technique_geometries(technique_geometries, node, parent_x=0, parent_y=0):
    """
    Iterate over an XML tree and create _TechniqueGeometry objects that represent each
    <g> element for a technique.

    SVG doesn't have any concept of depth or z-index. It draws all items in document
    order. So in order to draw an overlay, the overlay must come after all elements that
    it intersects. The straightforward way to do this is to append the element to the
    end of the document, but then it won't be translated properly since it's no longer
    nested inside the same <g> as the target element. So this method traverses the tree
    and keeps track of the sum of translations for each element, then writes a new
    technique <g> back to `technique_geometries` with the technique ID as key and its
    translation as values.

    :param ElementTree node:
    :rtype: _TechniqueGeometry
    """
    for child in node:
        child_x, child_y = parent_x, parent_y
        if transform := child.get("transform"):
            if match := TRANSLATE_RE.match(transform):
                child_x += float(match.group(1))
                child_y += float(match.group(2))
            else:
                raise ValueError(
                    f'Unknown SVG transform: "{transform}" (was this SVG created by ATT&CK Navigator?)'
                )
        if match := TECHNIQUE_CLASS_RE.match(child.get("class", "")):
            technique_id = match.group(2)
            try:
                rect = child.find("{http://www.w3.org/2000/svg}rect")
                width = float(rect.attrib["width"])
                height = float(rect.attrib["height"])
            except (KeyError, TypeError):
                raise ValueError(
                    f"Cannot find a <rect> with width and height attributes. The SVG may be a different format than this script requires. (The regex match was: {match})"
                )
            technique_geometries[technique_id] = _TechniqueGeometry(
                child_x, child_y, width, height
            )
        _enumerate_technique_geometries(technique_geometries, child, child_x, child_y)


def _create_technique_overlay(technique_id, technique_geometry):
    """
    Create the overlay object (a red ellipse).

    :param str technique_id:
    :param _TechniqueGeometry technique_geometry:
    :returns: a new <g> element
    """
    x, y, width, height = technique_geometry
    overlay_element = ElementTree.Element(
        "g", {"class": f"overlay {technique_id}", "transform": f"translate({x}, {y})"}
    )
    ellipse_el = ElementTree.Element(
        "ellipse",
        {
            "rx": str(width * ELLIPSE_SIZE_MULTIPLIER),
            "ry": str(height * ELLIPSE_SIZE_MULTIPLIER),
            "cx": str(width / 2),
            "cy": str(height / 2),
            "style": f"fill: none; stroke: {OVERLAY_COLOR}; stroke-width: {STROKE_WIDTH};",
        },
    )
    overlay_element.append(ellipse_el)
    return overlay_element


def _create_relationship_overlay(
    source_tid, source_geom, target_tid, target_geom, show_control_points
):
    """
    Create the overlay object (a red line with an arrowhead)

    :param str source_tid: source technique ID
    :param _TechniqueGeometry source_geom:
    :param str target_tid: target technique ID
    :param _TechniqueGeometry target_geom:
    :returns: a new <g> element
    """
    overlay_element = ElementTree.Element(
        "g",
        {
            "class": f"overlay {source_tid}-{target_tid}",
            "transform": f"translate({source_geom.x + source_geom.width / 2}, {source_geom.y + source_geom.height / 2})",
        },
    )

    # Compute the coordinates for the path.
    start_rx = source_geom.width * ELLIPSE_SIZE_MULTIPLIER
    start_ry = source_geom.height * ELLIPSE_SIZE_MULTIPLIER
    start_x = ARROW_OFFSET
    start_y = math.sqrt(start_ry**2 * (1 - start_x**2 / start_rx**2))

    target_rx = target_geom.width * ELLIPSE_SIZE_MULTIPLIER
    target_ry = target_geom.height * ELLIPSE_SIZE_MULTIPLIER
    target_dx = ARROW_OFFSET
    target_dy = math.sqrt(target_ry**2 * (1 - target_dx**2 / target_rx**2))
    target_x = target_geom.x - source_geom.x
    target_y = target_geom.y - source_geom.y

    distance = math.sqrt((target_x - start_x) ** 2 + (target_y - start_y) ** 2)
    control_dx = distance * 0.1
    control_dy = distance * 0.2
    # Make room for the arrow head. The slope is the same as control_dy / control_dx
    # so that the arrow points in the same direction as the line end.
    arrow_dx = math.sqrt((ARROWHEAD_WIDTH * 2) ** 2 / 5)
    arrow_dy = arrow_dx * 2

    if target_geom.x < source_geom.x:
        start_x = -start_x
        target_x += target_dx + arrow_dx
        control1_x = start_x - control_dx
        control2_x = target_x + control_dx
    else:
        target_x -= target_dx + arrow_dx
        control1_x = start_x + control_dx
        control2_x = target_x - control_dx

    if target_geom.y < source_geom.y:
        start_y = -start_y
        target_y += target_dy + arrow_dy
        control1_y = start_y - control_dy
        control2_y = target_y + control_dy
    else:
        target_y -= target_dy + arrow_dy
        control1_y = start_y + control_dy
        control2_y = target_y - control_dy

    if show_control_points:
        overlay_element.append(
            ElementTree.Element(
                "circle",
                {
                    "cx": str(start_x),
                    "cy": str(start_y),
                    "r": "1",
                    "style": "fill: green;",
                },
            )
        )

        overlay_element.append(
            ElementTree.Element(
                "circle",
                {
                    "cx": str(target_x),
                    "cy": str(target_y),
                    "r": "1",
                    "style": "fill: blue;",
                },
            )
        )

        overlay_element.append(
            ElementTree.Element(
                "circle",
                {
                    "cx": str(control1_x),
                    "cy": str(control1_y),
                    "r": "1",
                    "style": "fill: red;",
                },
            )
        )

        overlay_element.append(
            ElementTree.Element(
                "circle",
                {
                    "cx": str(control2_x),
                    "cy": str(control2_y),
                    "r": "1",
                    "style": "fill: orange;",
                },
            )
        )

    # Create the path.
    path_el = ElementTree.Element(
        "path",
        {
            "d": f"M {start_x} {start_y} C {control1_x} {control1_y} {control2_x} {control2_y} {target_x} {target_y}",
            "style": f"stroke: {OVERLAY_COLOR}; stroke-width: {STROKE_WIDTH}; fill: none;",
            "marker-end": "url(#attack-flow-arrowhead)",
        },
    )
    overlay_element.append(path_el)
    return overlay_element


def _get_arrowhead_marker():
    """
    Create a <marker> for drawing arrowheads on lines.

    :rtype: etree.Element
    """
    marker = ElementTree.Element(
        "marker",
        {
            "id": "attack-flow-arrowhead",
            "markerWidth": str(ARROWHEAD_WIDTH),
            "markerHeight": str(ARROWHEAD_HEIGHT),
            "refX": "0",
            "refY": str(ARROWHEAD_HEIGHT / 2),
            "orient": "auto",
        },
    )
    marker.append(
        ElementTree.Element(
            "polygon",
            {
                "points": f"0 0, {ARROWHEAD_WIDTH} {ARROWHEAD_HEIGHT / 2}, 0 {ARROWHEAD_HEIGHT}",
                "fill": OVERLAY_COLOR,
            },
        )
    )

    return marker
